workflows/rnaseq-ref-index/make_pre_mrna_fasta.R

```diff
@@ -0,0 +1,121 @@
+# to spliced and unspliced transcripts, then subsets the genome and gtf for 
+# only those particular regions, and finally outputs an expanded.fa, 
+# expanded.gtf, and expanded.tx2gene.tsv all corresponding to the genomic 
+# regions with spliced and unspliced transcripts.
+
+# Additionally this script creates a transcript2gene mapping file 
+# for use in salmon alevin, and a output a list of mitochondrial 
+# genes for use in QC. 
+
+# load needed packages
+library(Biostrings)
+library(BSgenome)
+library(eisaR)
+library(GenomicFeatures)
+library(magrittr)
+library(optparse)
+library(tidyverse)
+
+# Set up optparse options
+option_list <- list(
+  make_option(
+    opt_str = c("-f", "--gtf"),
+    type = "character",
+    default = "annotation/Homo_sapiens.GRCh38.103.gtf.gz",
+    help = "File path for input gtf file",
+  ),
+  make_option(
+    opt_str = c("-g", "--genome"),
+    type = "character",
+    default = "fasta/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz",
+    help = "File path for reference fasta file",
+  ), 
+  make_option(
+    opt_str = c("-O", "--output_dir"),
+    type = "character",
+    default = "annotation",
+    help = "Directory to write output files",
+  )
+)
+
+# Parse options
+opt <- parse_args(OptionParser(option_list = option_list))
+
+# Make output paths 
+expanded_fasta <- file.path(opt$output_dir, 
+                            "Homo_sapiens.ensembl.103.expanded.fa")
+expanded_gtf <- file.path(opt$output_dir, 
+                          "Homo_sapiens.ensembl.103.expanded.gtf")
+expanded_tx2gene <- file.path(opt$output_dir, 
+                              "Homo_sapiens.ensembl.103.tx2gene_expanded.tsv")
+tx2_gene_out <- file.path(opt$output_dir, 
+                      "Homo_sapiens.ensembl.103.tx2gene.tsv")
+mito_out <- file.path(opt$output_dir,
+                      "Homo_sapiens.ensembl.103.mitogenes.txt")
+
+# Check for output directory 
+if (!dir.exists(opt$output_dir)) {
+  dir.create(opt$output_dir)
+}
+
+# extract GRanges object containing genomic coordinates of each 
+# annotated transcript - both spliced and unspliced transcripts
+grl <- eisaR::getFeatureRanges(
+  gtf = opt$gtf,
+  featureType = c("spliced", "unspliced"), 
+  verbose = TRUE
+)
+
+# load in primary genome
+genome <- Biostrings::readDNAStringSet(opt$genome)
+names(genome) <- sapply(strsplit(names(genome), " "), .subset, 1)
+
+# extract unspliced and spliced sequences
+# genomic regions defined above
+seqs <- GenomicFeatures::extractTranscriptSeqs(
+  x = genome, 
+  transcripts = grl
+)
+
+# write unspliced and spliced sequences to fasta file
+Biostrings::writeXStringSet(
+  seqs, filepath = expanded_fasta
+)
+
+# write the associated annotations to gtf 
+eisaR::exportToGtf(
+  grl, 
+  filepath = expanded_gtf
+)
+
+# create text file mapping transcript and intron identifiers to 
+# corresponding gene identifiers
+# first make Tx2Gene for all spliced and unspliced sequences
+full_tx2gene <- eisaR::getTx2Gene(
```

Comment:
So I am thinking about this file and wondering if it is what we are really going to want, or if we need to modify it before use (and this may affect the gtf too?). This is a question about downstream processing, really. Do we want the genes in the tx2gene to be separated by spliced and unspliced, or only include the spliced (original) ids? I kind of think we do want the original only, as otherwise we would run into some problems where reads would be equally mapped to spliced & unspliced, then we have to combine them later. 

The simplest case is to think about a gene with one transcript (maybe one with no introns). After this procedure, we will have 2 "transcripts": `ENST000` and `ENST000-U` and a bunch of reads with counts that are getting split between them (we better not use any setting that throws out multimapped reads!). Those counts will then be split between gene `ENSG000` and `ENSG000-U`, but they should all be assigned to `ENSG000`, unless we are trying to keep them all separate. If we were doing RNA-velocity, we would want to keep track, of course, but we aren't at least not at this point, so does it make sense to change all the gene ids to drop the `-U`?

Alternatively, we might want to go back to the version with introns separately and then join intron counts in later analysis? How did you handle this in the past? How does CellRanger handle it?